pkg/logging/signoz.go

```diff
@@ -0,0 +1,282 @@
+	body, _ := zerologEntry[zerolog.MessageFieldName].(string)
+
+	attributes := make(map[string]string)
+	attributes["version"] = version.GetVersion()
+
+	for k, v := range zerologEntry {
+		if k == zerolog.TimestampFieldName || k == zerolog.LevelFieldName || k == zerolog.MessageFieldName {
+			continue
+		}
+		switch k {
+		case zerolog.CallerFieldName:
+			attributes["code.filepath"], _ = v.(string)
+		case zerolog.ErrorStackFieldName:
+			attributes["exception.stacktrace"], _ = v.(string)
+		default:
+			attributes[k] = fmt.Sprintf("%v", v)
+		}
+	}
+
+	logEntry := SigNozLogEntry{
+		Timestamp:      tsNano,
+		SeverityText:   severityText,
+		SeverityNumber: severityNumber,
+		Body:           body,
+		Attributes:     attributes,
+		Resources:      sw.resourceAttrs, // Use pre-configured resource attributes
+	}
+
+	sw.mu.Lock()
+	sw.logBuffer = append(sw.logBuffer, logEntry)
+	sw.mu.Unlock()
+
+	return len(p), nil
+}
+
+func (sw *SigNozJsonWriter) runSender() {
+	defer sw.wg.Done()
+	for {
+		select {
+		case <-sw.ticker.C:
+			sw.flushBuffer()
+		case <-sw.doneChan:
+			sw.ticker.Stop()
+			sw.flushBuffer() // Final flush
+			return
+		}
+	}
+}
+
+func (sw *SigNozJsonWriter) flushBuffer() {
+	sw.mu.Lock()
+	if len(sw.logBuffer) == 0 {
+		sw.mu.Unlock()
+		return
+	}
+	batchToSend := make([]SigNozLogEntry, len(sw.logBuffer))
+	copy(batchToSend, sw.logBuffer)
+	sw.logBuffer = make([]SigNozLogEntry, 0, sw.maxBatchSize) // Clear buffer
+	sw.mu.Unlock()
+
+	if len(batchToSend) == 0 {
+		return
+	}
+
+	jsonData, err := json.Marshal(batchToSend)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "SigNozJsonWriter: Error marshalling log batch: %v\n", err)
+		return
+	}
+
+	req, err := http.NewRequest("POST", sw.endpoint, bytes.NewBuffer(jsonData))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "SigNozJsonWriter: Error creating HTTP request: %v\n", err)
+		return
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("signoz-access-token", sw.accessToken)
+
+	resp, err := sw.httpClient.Do(req)
+	if err != nil {
+		// TODO NR - add backoff?
+		fmt.Fprintf(os.Stderr, "SigNozJsonWriter: Error sending log batch to SigNoz: %v\n", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 {
+		bodyBytes, _ := io.ReadAll(resp.Body)
+		fmt.Fprintf(os.Stderr, "SigNozJsonWriter: SigNoz returned non-2xx status: %d. Response: %s\n", resp.StatusCode, string(bodyBytes))
+	} else {
+	}
+}
+
+func (sw *SigNozJsonWriter) Close() error {
+	if sw.endpoint == "" || sw.accessToken == "" { // If writer was disabled
+		return nil
+	}
+	fmt.Fprintln(os.Stdout, "SigNozJsonWriter: Close called, attempting to flush remaining logs...")
+	close(sw.doneChan)
+	sw.wg.Wait()
+	fmt.Fprintln(os.Stdout, "SigNozJsonWriter: Closed.")
+	return nil
+}
+
+func CloseLoggers() {
+	log.Info().Msg("Closing loggers...")
+	if globalSigNozWriter != nil {
+		globalSigNozWriter.Close()
+	}
+	log.Info().Msg("Loggers closed.")
+}
+
+func getOtelCreds() (string, string, error) {
```

Comment:
This function exists in `metrics.otel`. Suggest moving `metrics` package to `/pkg` and importing from there.